src/crimson/os/seastore/segment_manager.h

```diff
@@ -159,10 +169,7 @@ class SegmentManager : public Device {
 
   /* Methods for discovering device geometry, segmentid set, etc */
   virtual seastore_off_t get_segment_size() const = 0;
-  virtual device_segment_id_t get_num_segments() const {
-    ceph_assert(get_size() % get_segment_size() == 0);
-    return ((device_segment_id_t)(get_size() / get_segment_size()));
-  }
+  virtual device_segment_id_t get_num_segments() const = 0;
```

Comment:
I think it should be always true/consistent that
```c++
get_size() == get_segment_size() * get_num_segment()
```
as of SegmentManager.

This means, only 2 of the 3 methods needs to be virtual in order to enforce this assumption:
```c++
device_segment_id_t get_num_segments() const {
  ceph_assert(get_size() % get_segment_size() == 0);
  return ((device_segment_id_t)(get_size() / get_segment_size()));
}
```

Otherwise, I think there will be surprises to the existing users of SegmentManager.

Reply:
`get_size() == get_segment_size() * get_num_segment()`
is true/consistent even now.

In the BlockSegmentManager implementation, only the starting 4K (plus another 8K, iirc) is used to store super block and remaining remaining space of entire device is used as segments.

For ZNS, first zone is fully reserved to store the super block as we cannot reuse this zone because resetting it would lose the super block.

So which of the 2 methods are you referring to here that needs to be virtual ?

Reply:
> `get_size() == get_segment_size() * get_num_segment()` is true/consistent even now.

I think not. After this PR, `get_num_segment()` will return `superblock.segments` in `BlockSegmentManager`, see `make_superblock()`:
```
size_t segments = (size - first_seg_off) / config_segment_size;
```
Which will not be equal to `get_size() / get_segment_size()`, which is `size / config_segment_size`.

However, it shows that the original version will return num-segments more than available in `BlockSegmentManager`, which was not correct I think.

Reply:
> So which of the 2 methods are you referring to here that needs to be virtual ? Is it in block.h and ephemeral.h ?

For the first virtual method in the Device class (device.h), let's rename it to:
```c++
virtual std::size_t get_available_size() const = 0;
```

The second virtual method is in SegmentManager class (segment_manager.h), keep it unchanged:
```c++
virtual seastore_off_t get_segment_size() const = 0;
```

Let's convert the third method to non-virtual in SegmentManager class (segment_manager.h):
```c++
// returns the number of available segments
device_segment_id_t get_num_segments() const {
  ceph_assert(get_available_size() % get_segment_size() == 0);
  return ((device_segment_id_t)(get_available_size() / get_segment_size()));
}
```
Does it make sense to you?

Reply:
Ohh. Yes. You are right. (I was thinking of zns.)

So for BlockSegmentManager,  in superblock,  we can set the size as `size = size  - first_segment_offset;`in block implementation and also update the number of segments as `segments = (size - first_segment_offset) / config_segment_size`. This would not use some of the media at the end of the disk roughly equivalent to or less than a segment size.

With above changes, get_num_segments() and get_size() would be returning consistent values.

Reply:
> > So which of the 2 methods are you referring to here that needs to be virtual ? Is it in block.h and ephemeral.h ?
> 
> For the first virtual method in the Device class (device.h), let's rename it to:
> 
> ```c++
> virtual std::size_t get_available_size() const = 0;
> ```
> 
> The second virtual method is in SegmentManager class (segment_manager.h), keep it unchanged:
> 
> ```c++
> virtual seastore_off_t get_segment_size() const = 0;
> ```
> 
> Let's convert the third method to non-virtual in SegmentManager class (segment_manager.h):
> 
> ```c++
> // returns the number of available segments
> device_segment_id_t get_num_segments() const {
>   ceph_assert(get_available_size() % get_segment_size() == 0);
>   return ((device_segment_id_t)(get_available_size() / get_segment_size()));
> }
> ```
> 
> Does it make sense to you?

Replied before refreshing this

Reply:
Please let me know if I got this correct.
1. rename get_size() to get_available_size() in device.h and in all subsequent places in segment_manager implementations, this will return the available_size in zns (zone-cap * num-of-segments) and in block implementation (size - first_segment_offset).
2. remove the patch `crimson/zns: Make get_num_segments() virtual`. So it makes get_num_segments as non-virtual .
3. use get_available_size() in get_num_segments().

Though I am not sure how it would fix the issue where more segments are reported than available, in block implementation, mentioned in your earlier comment.  So I think it would still need to be fixed as per my previous comment https://github.com/ceph/ceph/pull/47518#discussion_r957190801

Reply:
> Please let me know if I got this correct.

Correct.

> Though I am not sure how it would fix the issue where more segments are reported than available, in block implementation ...

As long as the implementation of `Device::get_available_size()` is correct --- the calculation looks correct in your comments --- the num segments won't be more than available. Let's see how it works.